### ISSUE TITLE
MFC-8440 Works Using MFC-8860DN at 600dpi (1200dpi has problems)

### ISSUE THREAD

**Issue by the-burrito-triangle** [pdewacht/brlaser#208]
The MFC-8440 prints fine at 600dpi using the MFC-8860DN PPD / brlaser cups filter. However, using 1200dpi produces artifacts along the left-hand-side margin (randomly shifted lines/blocks of content) and prints too large (looks like it just scales things larger rather than increasing the print DPI). Also, the default margins in the MFC-8860DN PPD are a bit extreme: 8 dots. While this region is printable (having lower toner density on the LHS with my printer), it is outside the default 18, 12 dots (horizontal and vertical margins) of the stock Brother PPD provided by the Brother OEM driver.

It would be nice to get away from the ancient 32bit Brother print driver for Linux. I think the brlaser CUPS filter just needs to add quirks for 1200dpi printing on older laser printers and it can likely support all Brother laser printers with slightly modified PPDs for the vertical and horizontal margins.

% sudo lpinfo --include-schemes usb -l -v

Device: uri = usb://Brother/MFC-8440?serial=000A5J443873
        class = direct
        info = Brother MFC-8440
        make-and-model = Brother MFC-8440
        device-id = MFG:Brother;CMD:HBP,PJL,PCL,PCLXL;MDL:MFC-8440;CLS:PRINTER;
        location =

**Comment by QORTEC** [comment]
Thank you for your report, your printer has been added to my pull request.

**Comment by the-burrito-triangle** [comment]
Thanks, QORTEC.

If it helps, the MFC-8440 uses the "1200HQ" mode (2400x600 DPI). Maybe this is why setting 1200DPI didn't work (MCF-8860DN might use the 1200x1200 mode)? If that could be fixed, I'd be a happy dude.

https://github.com/Owl-Maintain/brlaser/issues/19 

Also, from past research, I believe the following models should all work under the same print settings (1200HQ, etc.):
MFC-8440
MFC-8640D
MFC-8840D
MFC-8840DN

**Comment by QORTEC** [comment]
Thanks for the information. I've made a note to look into this resolution issue.

Can you please scan, or take a picture of the page so that I can see what your talking about?

Do you know if you are using the upstream `pdewacht` or the  `Owl-Maintain` fork of `brlaser`?

If you are using the  `Owl-Maintain` fork, there was a bug regarding duplex & tumble code, which caused errors when printing. But it sounds like issue you are having is different.

**Comment by the-burrito-triangle** [comment]
I was using the brlaser package provided by Fedora 39, which looks to be from the Owl-Maintain fork:
```
$ dnf info printer-driver-brlaser.x86_64
Name         : printer-driver-brlaser
Version      : 6.2.6
Release      : 1.fc39
Architecture : x86_64
Size         : 38 k
Source       : printer-driver-brlaser-6.2.6-1.fc39.src.rpm
Repository   : fedora
Summary      : Brother laser printer driver
URL          : https://github.com/Owl-Maintain/brlaser
License      : GPL-2.0-or-later
Description  : brlaser is a CUPS driver for Brother laser printers.
             : 
             : Although most Brother printers support a standard printer language
             : such as PCL or PostScript, not all do. If you have a monochrome
             : Brother laser printer (or multi-function device) and the other open
             : source drivers don't work, this one might help.
             : 
             : For a detailed list of supported printers, please refer to
             : https://github.com/Owl-Maintain/brlaser
```
I've since switched to using the hpcups driver with LaserJet 4300 emulation over the LAN (http://192.168.xxx.xxx:631/ipp/port1?version=1.0), due to issues with manually adjusting the print margins with brlaser (I forget exactly, but it seemed that either the left or right margin wouldn't adjust correctly). But I can try printing again with brlaser at 1200dpi and scan in the result. I've been meaning to upgrade to Fedora 40 or 41 (with associated newer brlaser packages), but GNOME 46 and up have issues with client side cursor scaling and I've not gotten around to seeing if they've fixed it yet. Maybe with Fedora 42 they'll have it fixed... Who knows.

**Comment by QORTEC** [comment]
> If that could be fixed, I'd be a happy dude.

@the-burrito-triangle I'm attempting to resolve the issue, and have a [`1200dpi`](/Owl-Maintain/brlaser/tree/1200dpi) branch on `Owl-Maintain/brlaser` that has code that more accurately mirrors how the official drivers handle 1200 dpi.

If you would like to test the new code, select the `A Owl‑Maintain/brlaser Test Driver Duplex` printer model; since that model has all the print resolutions are available.

**Comment by the-burrito-triangle** [comment]
Will do, thanks!

**Comment by the-burrito-triangle** [comment]
I'm getting some errors when building this. Should I try testing anyway?
```
$ cmake .
-- The CXX compiler identification is GNU 15.2.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No build type selected, default to RelWithDebInfo
-- Performing Test COMPILER_SUPPORT__Wall
-- Performing Test COMPILER_SUPPORT__Wall - Success
-- Performing Test COMPILER_SUPPORT__Wno_missing_braces
-- Performing Test COMPILER_SUPPORT__Wno_missing_braces - Success
-- Performing Test COMPILER_SUPPORT__Wdate_time
-- Performing Test COMPILER_SUPPORT__Wdate_time - Success
-- Performing Test COMPILER_SUPPORT__fstack_protector_strong
-- Performing Test COMPILER_SUPPORT__fstack_protector_strong - Success
-- Performing Test COMPILER_SUPPORT__Wformat
-- Performing Test COMPILER_SUPPORT__Wformat - Success
-- Performing Test COMPILER_SUPPORT__Wformat__Werror_format_security
-- Performing Test COMPILER_SUPPORT__Wformat__Werror_format_security - Success
-- Performing Test COMPILER_SUPPORT__D_FORTIFY_SOURCE_2
-- Performing Test COMPILER_SUPPORT__D_FORTIFY_SOURCE_2 - Success
-- Looking for C++ include cups/raster.h
-- Looking for C++ include cups/raster.h - found
-- Configuring done (0.7s)
-- Generating done (0.0s)
-- Build files have been written to: /mnt/ramdisk/brlaser-1200dpi
$ make
[  6%] Building CXX object CMakeFiles/rastertobrlaser.dir/src/main.cc.o
[ 12%] Building CXX object CMakeFiles/rastertobrlaser.dir/src/job.cc.o
[ 18%] Building CXX object CMakeFiles/rastertobrlaser.dir/src/line.cc.o
[ 25%] Building CXX object CMakeFiles/rastertobrlaser.dir/src/debug.cc.o
[ 31%] Linking CXX executable rastertobrlaser
[ 31%] Built target rastertobrlaser
[ 37%] Building CXX object CMakeFiles/test_lest.dir/test/test_lest.cc.o
[ 43%] Linking CXX executable test_lest
[ 43%] Built target test_lest
[ 50%] Building CXX object CMakeFiles/test_line.dir/test/test_line.cc.o
In file included from /usr/include/c++/15.2.1/vector:67,
                 from /mnt/ramdisk/brlaser-1200dpi/test/test_line.cc:21:
In function ‘_ForwardIterator std::uninitialized_fill_n(_ForwardIterator, _Size, const _Tp&) [with _ForwardIterator = unsigned char*; _Size = long unsigned int; _Tp = unsigned char]’,
    inlined from ‘_ForwardIterator std::__uninitialized_fill_n_a(_ForwardIterator, _Size, const _Tp&, allocator<_Tp>&) [with _ForwardIterator = unsigned char*; _Size = long unsigned int; _Tp = unsigned char; _Tp2 = unsigned char]’ at /usr/include/c++/15.2.1/bits/stl_uninitialized.h:727:39,
    inlined from ‘void std::vector<_Tp, _Alloc>::_M_fill_insert(iterator, size_type, const value_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’ at /usr/include/c++/15.2.1/bits/vector.tcc:686:36,
    inlined from ‘std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::insert(const_iterator, size_type, const value_type&) [with _Tp = unsigned char; _Alloc = std::allocator<unsigned char>]’ at /usr/include/c++/15.2.1/bits/stl_vector.h:1575:16,
    inlined from ‘<lambda()>’ at /mnt/ramdisk/brlaser-1200dpi/test/test_line.cc:128:16:
/usr/include/c++/15.2.1/bits/stl_uninitialized.h:553:39: warning: ‘void* __builtin_memset(void*, int, long unsigned int)’ offset [3, 514] is out of the bounds [0, 3] [-Warray-bounds=]
  553 |                       __builtin_memset(__dest, (unsigned char)__x, __n);
      |                       ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 56%] Building CXX object CMakeFiles/test_line.dir/src/line.cc.o
[ 62%] Linking CXX executable test_line
[ 62%] Built target test_line
[ 68%] Building CXX object CMakeFiles/test_block.dir/test/test_block.cc.o
[ 75%] Linking CXX executable test_block
[ 75%] Built target test_block
[ 81%] Building CXX object CMakeFiles/test_job.dir/test/test_job.cc.o
[ 87%] Building CXX object CMakeFiles/test_job.dir/src/job.cc.o
[ 93%] Building CXX object CMakeFiles/test_job.dir/src/line.cc.o
[100%] Linking CXX executable test_job
[100%] Built target test_job
```

**Comment by QORTEC** [comment]
Yes, I also got a similar warning when compiling, I haven't had time to look into the cause.
The driver should still work, so feel free to test it.

**Comment by the-burrito-triangle** [comment]
Thanks. I'll get it tested here soon.

**Comment by the-burrito-triangle** [comment]
Do I have to use USB with brlaser? I currently have my printer setup for use over the LAN.

**Comment by QORTEC** [comment]
Yes, the driver supports network printing.

**Comment by the-burrito-triangle** [comment]
Okay, when I go to add a new printer and put the IPP (internet printing protocol) address in which printer should I pick?

<img width="1008" height="1028" alt="Image" src="https://github.com/user-attachments/assets/74a79ae6-1a54-4e60-8b4e-e68a2a85639e" />

**Comment by the-burrito-triangle** [comment]
I guess I could use the HP Laserjet and manually edit the .ppd. (I've had to edit the .ppd when testing previously, so it's not that much trouble to manually configure the .ppd to use the `rastertobrlaser` filter.)

**Comment by the-burrito-triangle** [comment]
Interesting, it looks like I can just pick whatever, go to "Printer Details" and then click "Select from Database" select "Brother" and then pick the right one from the drop down. It's been awhile. I figured I could just type the printer's address and select the brlaser option from there.

**Comment by the-burrito-triangle** [comment]
Alright, I selected the `A Owl‑Maintain/brlaser Test Driver` for non-duplex printers (since mine can't do duplex) but there are multiple 1200dpi options. Which one is for the 2400x600 mode?

<img width="1590" height="1128" alt="Image" src="https://github.com/user-attachments/assets/92070d16-1b3a-4867-af31-83093210dc30" />

**Comment by QORTEC** [comment]
Sorry I got distracted investigating and creating a bug report on the compiler warning.

Brother uses multiple 1200dpi modes:
1200dpi = 1200x1200dpi
HD1200A = 1200x600dpi
HD1200B = 2400x600dpi

You'll probably want the last one, but the other ones may also work.

**Comment by the-burrito-triangle** [comment]
Thanks, is there a specific test image I should try printing? Maybe a checker board mesh or something?

**Comment by QORTEC** [comment]
Doesn't matter, anything that wasn't working at 1200dpi is fine.
I just want confirmation from people who had issues printing 1200dpi that the issue seems to be fixed.

**Comment by QORTEC** [comment]
I never had any issues at 1200dpi, so I cant really thoroughly debug and test to make sure its squashed.

**Comment by the-burrito-triangle** [comment]
Yeah, the 1200B is still over sized. I'll try the others and see if one of them works right.

**Comment by QORTEC** [comment]
Just to double check is you issue similar to Owl-Maintain/brlaser#38?

**Comment by the-burrito-triangle** [comment]
All three 1200 dpi modes print identically: makes the image too large and the printer simply crops it and prints what fits.

When using 600dpi the image prints fine but the left and right margins are not equal. The LHS has more space than the RHS.

**Comment by the-burrito-triangle** [comment]
That does appear to be the same issue I'm seeing (image is scaled up and cropped when printed).

I'll scan the prints so you can see what's happening.

**Comment by QORTEC** [comment]
Thank you for testing, It looks like this issue still needs to be investigated farther.

**Comment by the-burrito-triangle** [comment]
My scans aren't great since simple-scan changed my settings after updating and I couldn't be bothered to fine tune the brightness and contrast again. But, this shows the same behavior as https://github.com/Owl-Maintain/brlaser/issues/38:

Original image:
<img width="2550" height="3300" alt="Image" src="https://github.com/user-attachments/assets/295da8f9-14dc-483f-8c39-1b2dc7bde304" />

600 DPI
<img width="2496" height="3298" alt="Image" src="https://github.com/user-attachments/assets/d1afb437-c017-4da7-8242-a351e8aa9bdd" />

HD1200B (HD1200A and 1200 DPI also print the same with the MFC-8440)
<img width="2496" height="3298" alt="Image" src="https://github.com/user-attachments/assets/7278da03-a32d-461c-acdb-5fb4c0e23f4e" />

**Comment by the-burrito-triangle** [comment]
Hey, I was wondering if using Wireshark to capture the output of the 32bit Brother driver with HQ1200 mode would help here?

**Comment by QORTEC** [comment]
More information would definitely be better; being able to confirm what commands are being sent by each driver should narrow down this issue.
